src/modm/platform/clock/sam_pmc/clockgen.hpp.in

```diff
@@ -105,15 +120,26 @@ class ClockGen {
 	 */
 	static void enableExternal32Khz(bool crystal_bypass = false);
 
+	// TODO: template?
+	static void enableInternal12M(MainInternalFreq freq);
+
 	template<uint32_t multiplier>
 	static void enablePllA(uint32_t wait_cycles = 50);
 
+%% if target["family"] in ["g"]
 	template<uint32_t multiplier>
 	static void enablePllB(uint32_t wait_cycles = 50);
+%% endif
+
+%% if target["family"] in ["v"]
+	template<uint32_t multiplier>
+	static void enableUPll(uint32_t wait_cycles = 50);
+%% endif
 
 	template <MasterClkSource src, MasterClkPrescaler pres>
 	static void selectMasterClk();
 
+	// TODO: does it make sense without enableInternal12M?
 	static void setMainInternalFreq(MainInternalFreq freq);
```

Comment:
@mcbridejc what is this for?
- I don't see this used anywhere
- The implementation seems incomplete (it doesn't wait for `MOSCRCS`)
- I think it's useful only inside enableInternal12M

I would suggest removing this method and moving it's body to enableInternal12M.

Reply:
It's for an application to switch the default 8MHz clock to 16 or 24MHz after boot, but I don't think there's any harm in enabling it if it's already enabled, so replacing that function with one which means "enable and set frequency" instead of just "set frequency" is good to me.

I think you should have three methods to setup the main clock:
`enableMainInternal(MainInternalFreq freq)`
`enableMainExtermal(bool bypass)`
`selectMainClockSource(MainClockSource src)`

```
enum class MainClockSource {
  Internal,
  External
}
```

I'm suggesting to separate selectMainClkSource from the enable functions, because It looks like on the samv it is possible to enable the external crystal(for the UPLL) while using the internal RC for main clock.

Reply:
Done. I didn't implement enableMainExternal intentionally, I don't have a way to test it yet.